### PR TITLE
Gsdx alpha test improvement DX

### PR DESCRIPTION
**Pull request by FlatOutPS2**
Port for the DX renderers of the alpha test improvement PR for OGL(#1514) created
by gregory38.

Tested and confirmed to work the same as OGL with both DX9 and DX11, fixing text issues for the Burnout games, Kengo and Ridge Racer V.

### CODE REVIEW

File: plugins/GSdx/GSRendererDX.cpp [hunk 1]
```diff
@@ -48,15 +48,176 @@ GSRendererDX::~GSRendererDX()
...
+	
+	// Minor optimization of a corner case (it allow to better emulate some alpha test effects)
+	if (om_dssel.ztst == ZTST_GEQUAL && m_vt.m_min.p.z >= max_z_flt) {
+#ifdef _DEBUG
+		fprintf("Optimize Z test GEQUAL to ALWAYS (%s)", psm_str(context->ZBUF.PSM));
+#endif
+		om_dssel.ztst = ZTST_ALWAYS;
+	}
+
+	if (m_fba)
```
Comment (gregory38) [inline plugins/GSdx/GSRendererDX.cpp]: Why did you put FBA & blending emulation too ? The function is called Z buffer emulation aka depth ;)

**Comment by FlatOutPS2** [comment]
> Why did you put FBA & blending emulation too ? The function is called Z buffer emulation aka depth ;)

I think it was because of a regression in an early version of the port. I completely forgot about it being there. :p

File: plugins/GSdx/GSRendererDX.cpp (hunk 1 above)
Comment (FlatOutPS2) [inline plugins/GSdx/GSRendererDX.cpp]: > Why did you put FBA & blending emulation too ? The function is called Z buffer emulation aka depth ;)

I think it was because of a regression in an early version of the port. I completely forgot about it being there. :p

**Comment by MrCK1** [comment]
I guess I'll test Vice City later. I'm not expecting anything different than the half-baked results I got in OGL

**Comment by FlatOutPS2** [comment]
If it's not fixed in the OGL version, it won't be any better in this. It's a port, not an upgrade. :D

**Comment by gregory38** [comment]
What a deception :p

Care to share a link to a gs dump. Current implementation is rather limited to a special case. (Depth test can be executed in an out of order). At least I can tell if it is the same bug or another one.

**Comment by FlatOutPS2** [comment]
> Care to share a link to a gs dump. Current implementation is rather limited to a special case. (Depth test can be executed in an out of order). At least I can tell if it is the same bug or another one.

I already asked for one in the issue he linked in your PR: #1329

**Comment by gregory38** [comment]
I have some pending change for ate.
* a case that wasn't correct
* a 2nd optimization for vice city that disable alpha test completely in some cases.
* And I'm looking at the remaining vice city issue (which is very close)

**Comment by gregory38** [comment]
Sorry, you need to update it again.

**Comment by FlatOutPS2** [comment]
That fixes the map in GTA. There's one more remaining issue that was fixed when I hack-fixed Burnout games, but wasn't touched by any of the changes in this PR, the dancing girl in Ridge Racer V's main menu is still split in half, left side is correct, depth of right side is wrong.

**Comment by gregory38** [comment]
```
the dancing girl in Ridge Racer V's main menu is still split in half, left side is correct, depth of right side is wrong.
```
Half screen issue feels like a texture cache issue rather than a wrong alpha test per se (I would say a wrong depth buffer content).

**Comment by FlatOutPS2** [comment]
> Half screen issue feels like a texture cache issue rather than a wrong alpha test per se (I would say a wrong depth buffer content).

It's not. It's not fixed by turning off texture cache and it was fixed by an alpha test modification.

**Comment by gregory38** [comment]
```
It's not. It's not fixed by turning off texture cache and it was fixed by an alpha test modification.
```
I don't think depth buffer is properly read when you disable the texture cache. Anyway, give me a gs link, if it is an alpha test issue it should be easy to find the bad draw call.

**Comment by FlatOutPS2** [comment]
GS link: [Ridge Racer V.zip](https://github.com/PCSX2/pcsx2/files/420601/Ridge.Racer.V.zip)

**Comment by gregory38** [comment]
Did you try to replay your dump? It seems to be fine. However there is a single ATE draw call. It is done to draw the circle. And I can see a half screen delimitation effect. But your screenshot seems to be fine on that side. Remaining alpha test is skipped. 

Hum, however there is a destination alpha test draw call to draw the girl but it is actually a nop. I need to run the game (I have it).

**Comment by gregory38** [comment]
I understand the line. The depth buffer is cleared with line. I need to finish a hack to replace point/line with square/rectangle to avoid such aliasing. I'm trying to look at the half screen now. The depth buffer is wrong so ATE later is broken.

File: plugins/GSdx/GSRendererDX.cpp [hunk 2]
```diff
@@ -48,15 +48,139 @@ GSRendererDX::~GSRendererDX()
 {
 }
 
+void GSRendererDX::EmulateAtst(const int pass, const GSTextureCache::Source* tex)
+{
+	GSDeviceDX* dev = (GSDeviceDX*)m_dev;
```
Comment (gregory38) [inline plugins/GSdx/GSRendererDX.cpp]: For the record I do that in openGL because it allow the compiler to optimize virtual function call. OpenGL device is the lastest inheritance, so you know the function at compile time. On dx it is useless you still need to dispatch it at runtime to select either dx9 or dx11.

File: plugins/GSdx/GSRendererDX.cpp [hunk 3]
```diff
@@ -48,15 +48,139 @@ GSRendererDX::~GSRendererDX()
...
+		om_dssel.ztst = ZTST_ALWAYS;
+	}
+}
+
 void GSRendererDX::DrawPrims(GSTexture* rt, GSTexture* ds, GSTextureCache::Source* tex)
 {
-	GSDrawingEnvironment& env = m_env;
-	GSDrawingContext* context = m_context;
+	context = m_context;
+	env = m_env;
```
Comment (gregory38) [inline plugins/GSdx/GSRendererDX.cpp]: This code suck because it will a (useless) copy of the env object. On openGL I replaced all occurence to use `m_` prefix instead. It doesn't worth it, it just save a couple of chars.
You can only replace it on the newly created function. It would be a better idea to do it in a separate commit if you do it.

Comment (FlatOutPS2) [inline plugins/GSdx/GSRendererDX.cpp]: > You can only replace it on the newly created function. 

What do you mean by that?

Comment (gregory38) [inline plugins/GSdx/GSRendererDX.cpp]: I mean the only pupose of `env` is to write `env.foo` instead of `m_env.foo`. In short it is useless. If you use `env...` in the new created function for depth/alpha, you just need to call `m_env....` instead.

Comment (FlatOutPS2) [inline plugins/GSdx/GSRendererDX.cpp]: Right, but why can I only replace it on the newly created function?

Comment (gregory38) [inline plugins/GSdx/GSRendererDX.cpp]: What I mean is that you could do
```
depth () {
 m_env....
}
draw() {
  env = m_env;
 env.ff
```
Or you can replace it everywhere. Just more change and a bigger diff but you're done for the next time. It is up to you. Or you could use the env = m_env renaming everywhere. But don't store it in the object, it is a bad idea.

Comment (FlatOutPS2) [inline plugins/GSdx/GSRendererDX.cpp]: Okay. But what about m_context?

Comment (gregory38) [inline plugins/GSdx/GSRendererDX.cpp]: Same stuff. Look at the code in the openGL renderer. You can see that I directly use `m_context`

Comment (FlatOutPS2) [inline plugins/GSdx/GSRendererDX.cpp]: Yes, I already did. But in the original GSRendererDX code both m_context and context are used, so I wasn't sure if it was required or not.

Comment (gregory38) [inline plugins/GSdx/GSRendererDX.cpp]: Old code uses context, newer code use m_context (imported from openGL)

File: plugins/GSdx/GSRendererDX.cpp [hunk 4]
```diff
@@ -48,21 +48,143 @@ GSRendererDX::~GSRendererDX()
...
+	// than the buffer supports seems to be an error condition on the real GS, causing it to crash.
+	// We are probably receiving bad coordinates from VU1 in these cases.
+
+	if (om_dssel.ztst >= ZTST_ALWAYS && om_dssel.zwe && (context->ZBUF.PSM != PSM_PSMZ32)) {
+		if (m_vt.m_max.p.z > max_z) {
+			ASSERT(m_vt.m_min.p.z > max_z); // sfex capcom logo
+			// Fixme :Following conditional fixes some dialog frame in Wild Arms 3, but may not be what was intended.
+			if (m_vt.m_min.p.z > max_z) {
+#ifdef _DEBUG
+				fprintf("Bad Z size on %s buffers", psm_str(context->ZBUF.PSM));
```
Comment (gregory38) [inline plugins/GSdx/GSRendererDX.cpp]: Does it compile on debug build ? Don't you need either `stdout` or `stderr` in first paramerter (or use `printf`). And you might want to add a final `\n`

File: plugins/GSdx/GSRendererDX.cpp [hunk 5]
```diff
@@ -48,21 +48,143 @@ GSRendererDX::~GSRendererDX()
...
+				om_dssel.ztst = ZTST_ALWAYS;
+			}
+		}
+	}
+
+	GSVertex* v = &m_vertex.buff[0];
+	// Minor optimization of a corner case (it allow to better emulate some alpha test effects)
+	if (om_dssel.ztst == ZTST_GEQUAL && (m_vt.m_eq.xyzf & 0x8) && v[0].XYZ.Z == max_z) {
+#ifdef _DEBUG
+		fprintf("Optimize Z test GEQUAL to ALWAYS (%s)", psm_str(context->ZBUF.PSM));
```
Comment (gregory38) [inline plugins/GSdx/GSRendererDX.cpp]: same as above

File: plugins/GSdx/GSRendererDX.cpp [hunk 6]
```diff
@@ -358,30 +437,30 @@ void GSRendererDX::DrawPrims(GSTexture* rt, GSTexture* ds, GSTextureCache::Sourc
...
-				ps_sel.atst = 5; // >=
-			else
-				ps_sel.atst = 2; // <
-			ps_cb.FogColor_AREF.a = (float)0x80;
-		}
-		if (!(context->FBA.FBA && context->TEST.DATM == 1))
-			om_dssel.alpha_stencil = 1;
+	if (ate_skip) {
+#ifdef _DEBUG
+		fprintf("Alternate ATE handling: ate_skip");
```
Comment (gregory38) [inline plugins/GSdx/GSRendererDX.cpp]: See previous comment for those 2 printfs too.

**Comment by gregory38** [comment]
Is it ready to be merged ? TBH, I don't like the placement of  `{` around if/else but clang-format will take of it.

**Comment by FlatOutPS2** [comment]
In terms of it working as good as the OGL version it was ready to be merged when I created the PR.

Placement of `{` is based on your PR, what's used in nearby inconsistent code and possible VS trickery.
